### PR TITLE
Fix/551 Improve Google Client creation

### PR DESCRIPTION
**Pull request by friedger**
This PR fixes the Google client creation process and makes sure that the previous client is released when creating a new one.

In particular, the client is created depending on the signin flag.

UI has not changed.

To test sign in and signout using the settings activity.

### CODE REVIEW

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java [hunk 1]
```diff
@@ -172,15 +163,13 @@ public AchievementActionHandler getAchievementActionHandler() {
...
 
-        if (PrefUtils.isSignedIn(this) && !Utils.isEmulator()) {
+        if (PrefUtils.isSignedIn(this)) {
             switch (requestCode) {
                 case RC_SIGN_IN:
                     if (responseCode == RESULT_OK) {
                         // If the error resolution was successful we should continue
                         // processing errors.
                         mSignInProgress = STATE_SIGN_IN;
-                        App.getInstance().resetOrganizer();
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: Why don't we need this anymore?

File: app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java [hunk 2]
```diff
@@ -278,6 +280,19 @@ public boolean onPreferenceChange(Preference preference, Object o) {
         }
     }
 
+    private void createConnectedGoogleApiClient() {
+        mGoogleApiClient = GoogleApiClientFactory.createWith(getActivity().getApplicationContext());
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java]: Is there a better way to create GoogleApiClient inside a Fragment?

**Comment by tasomaniac** [comment]
LGTM. I have just 1 comment about organizer check. I want to download it and test it before merging.

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java (hunk 1 above)
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: It did not reset anything but just overwrote the organizerChecker with values from preferences.

The resetting is now properly covered for in checkOrganizer: https://github.com/gdg-x/frisbee/pull/560/files#diff-fd216eefb97ccc0dfae2b7ae369f752bR42

**Comment by tasomaniac** [comment]
Sorry for the really late reply. I've tested it on the device. It seems that 3rd bullet in the issue #551 is not fixed. 
If Arrow or Achievements clicked when the user is signed-out, we ask the user to signin with a dialog. Clicking SIGN-IN in that dialog does nothing.

**Comment by tasomaniac** [comment]
Another crash that I just got. Signin in the FirstStart is not working. 

screen shot 2016-03-06 at 20 11 49

File: app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java [hunk 3]
```diff
@@ -44,9 +45,7 @@
 
     @Override
     protected GoogleApiClient createGoogleApiClient() {
-        return new GoogleApiClient.Builder(this)
-            .addApi(AppInvite.API)
-            .build();
+        return GoogleApiClientFactory.createWithoutSignIn(getApplicationContext());
```
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java]: I am not sure if it is a good practice to create GoogleApiClient with ApplicationContext. 

GoogleApiClient is really tied to the Activity. We are recreating it on every Activity. That's how it needs to be done. So I am not sure if this makes sense.

**Comment by tasomaniac** [comment]
All the problems described in #551 is fixed now. But it need Code Review. Please review this and merge.

**Comment by tasomaniac** [comment]
It would be great if multiple people review this. @fnk0 @PareshMayani @rciovati

**Comment by friedger** [comment]
I have reviewed @tasomaniac commits. It would be nice to have a third person to merge this PR!

File: app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java (hunk 3 above)
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java]: If the client were not leaking the acitivity I am happy to use the activity.

I couldn't find different solution to resolve the leaking issue with the clients.

**Comment by stefanhoth** [comment]
@friedger if you comment on the commit we won't see the detail in the discussion here and it doesn't get closed automatically.

Having a look at the diff now.

File: app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java [hunk 4]
```diff
@@ -134,7 +133,7 @@ public void onConnectionSuspended(int i) {
     }
 
     @Override
-    public void onConnectionFailed(ConnectionResult connectionResult) {
+    public void onConnectionFailed(@NonNull ConnectionResult connectionResult) {
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java]: Is `@NonNull` a project convention? I usually would advise against it because this should be the default for defensive programming.

File: app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+import com.google.android.gms.appindexing.AppIndex;
+import com.google.android.gms.common.api.GoogleApiClient;
+import com.google.android.gms.drive.Drive;
+import com.google.android.gms.games.Games;
+import com.google.android.gms.plus.Plus;
+
+import org.gdg.frisbee.android.utils.PrefUtils;
+
+public final class GoogleApiClientFactory {
+    private GoogleApiClientFactory() { }
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: I'd prefer to put the static methods to the top of the class. Is there any convention in this project already?

File: app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java [hunk 6]
```diff
@@ -249,7 +249,7 @@ private String getChapterIdFromIntent(final Intent intent) {
         return null;
     }
 
-    private void fetchChapters() {
+    void fetchChapters() {
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java]: Why did the visibility change for all methods/members?

Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java]: They are accessed by anonymous classes, this change makes the fact obvious.

Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java]: This is a new lint issue in 2.1 preview

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java [hunk 7]
```diff
@@ -295,7 +280,7 @@ public void onConnectionSuspended(int i) {
     }
 
     @Override
-    public void onConnectionFailed(ConnectionResult result) {
+    public void onConnectionFailed(@NonNull ConnectionResult result) {
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: See https://github.com/gdg-x/frisbee/pull/560/files#r55929152 for the same question.

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java [hunk 8]
```diff
@@ -327,7 +312,7 @@ protected void recordStartPageView(Action viewAction) {
     protected ResultCallback<Status> appIndexApiCallback(final String label) {
         return new ResultCallback<Status>() {
             @Override
-            public void onResult(Status status) {
+            public void onResult(@NonNull Status status) {
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: See https://github.com/gdg-x/frisbee/pull/560/files#r55929152 for the same question.

File: app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java (hunk 5 above)
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: All conventions are defined in .editorconfig, so no.

File: app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java [hunk 9]
```diff
@@ -416,7 +416,7 @@ public void onNothingSelected(final AdapterView<?> parent) {
         );
     }
 
-    private void updateSelectionfor(final String newChapterId) {
+    void updateSelectionfor(final String newChapterId) {
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java]: Not your mistake but it should be `updateSelectionFor`

File: app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java (hunk 5 above)
Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: @stefanhoth Could you open an issue to discuss this convention?

Comment (friedger) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: out of scope for this PR, IMHO.

Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: @friedger Do you mean an issue to discuss/create a convention for the order of methods or for this specific case?

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java [hunk 10]
```diff
@@ -118,69 +108,86 @@ public GoogleApiClient getGoogleApiClient() {
...
+        if (!mGoogleApiClient.isConnectionCallbacksRegistered(this)) {
+            mGoogleApiClient.registerConnectionCallbacks(this);
+        }
+        if (!mGoogleApiClient.isConnectionFailedListenerRegistered(this)) {
+            mGoogleApiClient.registerConnectionFailedListener(this);
+        }
+    }
+
+    /**
+     * Create {@link GoogleApiClient}. This can be overridden to change the scope of the GoogleApiClient
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: :+1:

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java [hunk 11]
```diff
@@ -118,69 +108,86 @@ public GoogleApiClient getGoogleApiClient() {
...
+    protected void onSaveInstanceState(Bundle outState) {
+        super.onSaveInstanceState(outState);
+        outState.putInt(SAVED_PROGRESS, mSignInProgress);
     }
 
     @Override
     protected void onActivityResult(int requestCode, int responseCode, Intent intent) {
         super.onActivityResult(requestCode, responseCode, intent);
 
-        if (PrefUtils.isSignedIn(this) && !Utils.isEmulator()) {
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: Why is the Emulator check removed? What changed?

File: app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java (hunk 5 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: In Utility classes, private constructor is always at the top in all classes in Frisbee. So if you don't count that one the rest is all static :) We can definitely discuss this, private constructor can be moved to bottom.

File: app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java (hunk 4 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/activity/AppInviteDeepLinkActivity.java]: Unfortunately lint warns me to do it because the parent implementation have that. We don't put @NonNull otherwise

File: app/src/main/java/org/gdg/frisbee/android/common/GdgNavDrawerActivity.java [hunk 12]
```diff
@@ -75,7 +79,25 @@
     ImageView mDrawerImage;
     ImageView mDrawerUserPicture;
     private ActionBarDrawerToggle mDrawerToggle;
-    private MenuItem drawerItemToNavigateAfterSignIn = null;
+
+    int drawerItemIdToNavigateAfterSignIn = INVALID_ITEM_ID;
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgNavDrawerActivity.java]: Was there a specific reason that this is part of the PR? Feels like this is a separate change that blows up the diff.

File: app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java (hunk 6 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/chapter/MainActivity.java]: It's because when private methods, fields are accessed by anonymous classes, compiler generates unnecessary methods. This avoids that.

File: app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java (hunk 5 above)
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: @tasomaniac Well, I don't want to change how the default is. Thanks for the clarification.

File: app/src/main/java/org/gdg/frisbee/android/common/GdgNavDrawerActivity.java (hunk 12 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgNavDrawerActivity.java]: This was MenuItem before. I want this to survive orientation/config changes. Since MenuItem is not Parcelable, I cannot put it in Bundle. 
Maybe orientation change related problems with Sign In could be put in a different PR. Sorry about that.

File: app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java (hunk 11 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgActivity.java]: This was here since Emulators had problems with Google Play Services.

Right now, we are trying to make the app work even there is no GMS or user does not have Google+ Google Play Games account. The app will work on every situation. So there is no need for emulator check.

File: app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java (hunk 5 above)
Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/app/GoogleApiClientFactory.java]: Thank you for the review. We can definitely change that project wise.

File: app/src/main/java/org/gdg/frisbee/android/common/GdgNavDrawerActivity.java (hunk 12 above)
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/common/GdgNavDrawerActivity.java]: ok, just checking :smile:

File: app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java [hunk 13]
```diff
@@ -63,120 +65,119 @@
...
-                                            PrefUtils.setGcmSettings(
-                                                getActivity(),
-                                                true,
-                                                regId,
-                                                gcmRegistrationResponse.getNotificationKey()
-                                            );
-                                        }
-                                    });
-
-                                setHomeGdg(PrefUtils.getHomeChapterIdNotNull(getActivity()));
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java]: So much horrible code deleted. :+1:

File: app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java [hunk 14]
```diff
@@ -63,120 +65,119 @@
...
+        new Preference.OnPreferenceChangeListener() {
+            @Override
+            public boolean onPreferenceChange(Preference preference, Object newValue) {
+                final boolean enableGcm = (Boolean) newValue;
+
+                if (mGoogleApiClient.isConnected()) {
+                    mLoading.setVisibility(View.VISIBLE);
+                    mLoading.startAnimation(AnimationUtils.loadAnimation(getActivity(), R.anim.fade_in));
+
+                    new AsyncTask<Void, Void, Void>() {
```
Comment (stefanhoth) [inline app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java]: Could this move into it's own class? The readability is horrible.

Comment (tasomaniac) [inline app/src/main/java/org/gdg/frisbee/android/fragment/SettingsFragment.java]: I hate these. :( But it was there before. Out of this PRs scope.
